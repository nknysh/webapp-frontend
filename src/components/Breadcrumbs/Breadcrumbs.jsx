import React from 'react';
import { withRouter } from 'react-router-dom';
import { compose, map, equals } from 'ramda';
import hash from 'object-hash';

import { propTypes, defaultProps } from './Breadcrumbs.props';
import { StyledBreadcrumbs, Crumb, CrumbLink } from './Breadcrumbs.styles';

export const Breadcrumbs = ({ location: { pathname }, links, ...props }) => {
  // eslint-disable-next-line
  const renderLinks = ({ label, to, ...crumbProps }) => (
    <Crumb key={hash({ label: label && label.toString(), to })}>
      {to ? (
        <CrumbLink data-active={equals(pathname, to)} to={to} {...crumbProps}>
          {label}
        </CrumbLink>
      ) : (
        label
      )}
    </Crumb>
  );
  return <StyledBreadcrumbs {...props}>{map(renderLinks, links)}</StyledBreadcrumbs>;
};

Breadcrumbs.propTypes = propTypes;
Breadcrumbs.defaultProps = defaultProps;

export default compose(withRouter)(Breadcrumbs);
