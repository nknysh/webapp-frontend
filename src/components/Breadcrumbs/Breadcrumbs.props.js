import PropTypes from 'prop-types';

export const propTypes = {
  links: PropTypes.arrayOf(PropTypes.object).isRequired,
};

export const defaultProps = {
  links: [],
};
