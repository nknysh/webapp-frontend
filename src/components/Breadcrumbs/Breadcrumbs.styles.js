import styled, { css } from 'styled-components';
import { Link } from '@pure-escapes/webapp-ui-components';

import { theme } from 'styles';

export const StyledBreadcrumbs = styled.div`
  padding: ${theme.spacing.gutter}px 0;
  margin: ${theme.spacing.gutter * 2}px 0;
  display: flex;
  flex-wrap: wrap;
  align-items: center;
`;

export const Crumb = styled.div`
    padding: 0 ${theme.spacing.gutter * 2.5}px;

    a {
        padding: 0 ${theme.spacing.gutter / 2}px !important;
    }

    :not(span){
        line-height: 20px; !important;
    }

    :not(:first-child) {
        border-left: 1px solid ${theme.borders.default};
    }

    :first-child {
        padding-left: 0;
    }
`;

export const CrumbLink = styled(Link)`
  color: ${theme.palette.primary} !important;
  font-weight: ${theme.fonts.bold};
  text-transform: uppercase;
  font-size: ${theme.fonts.sizes.default}px;
  letter-spacing: ${theme.fonts.letterSpacing.medium}px;

  padding-top: 8px !important;

  ${({ ['data-active']: active }) =>
    active &&
    css`
      color: ${theme.palette.neutral} !important;
    `}
`;
