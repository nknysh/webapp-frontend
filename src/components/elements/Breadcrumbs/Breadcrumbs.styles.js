import styled, { css } from 'styled-components';

import Link from 'components/elements/Link';

import theme from 'styles/theme';

export const StyledBreadcrumbs = styled.div`
  padding: ${theme.gutter}px 0;
  margin: ${theme.gutter * 2}px 0;
  display: flex;
  flex-wrap: wrap;
  align-items: center;
`;

export const Crumb = styled.div`
    padding: 0 ${theme.gutter * 2.5}px;

    a {
        padding: 0 ${theme.gutter / 2}px !important;
    }

    :not(span){
        line-height: 20px; !important;
    }

    :not(:first-child) {
        border-left: 1px solid ${theme.borders.default};
    }

    :first-child {
        padding-left: 0;
    }
`;

export const CrumbLink = styled(Link)`
  color: ${theme.primary} !important;
  font-weight: ${theme.fonts.bold};
  text-transform: uppercase;
  font-size: ${theme.fonts.sizes.default}px;
  letter-spacing: ${theme.fonts.letterSpacing.medium}px;

  padding-top: 8px !important;

  ${({ ['data-active']: active }) =>
    active &&
    css`
      color: ${theme.neutral} !important;
    `}
`;
