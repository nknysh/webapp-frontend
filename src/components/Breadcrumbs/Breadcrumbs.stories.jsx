import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import { range } from 'ramda';

import { mapWithIndex } from 'utils';

import { storiesOf } from '@storybook/react';
import { withKnobs, number, text } from '@storybook/addon-knobs';

import Breadcrumbs from './Breadcrumbs';

storiesOf('Breadcrumbs', module)
  .addDecorator(withKnobs)
  .add('default', () => {
    const renderBreadcrumb = number => ({
      label: `Crumb ${number}`,
      to: `/${number}`,
    });

    return (
      <BrowserRouter>
        <Breadcrumbs
          location={{ pathname: text('Current path', '/1') }}
          links={mapWithIndex(
            renderBreadcrumb,
            range(
              0,
              number('Number of crumbs', 2, {
                range: true,
                min: 1,
                max: 5,
                step: 1,
              })
            )
          )}
        />
      </BrowserRouter>
    );
  });
