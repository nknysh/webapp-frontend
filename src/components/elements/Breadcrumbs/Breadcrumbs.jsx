import React from 'react';
import { withRouter } from 'react-router-dom';
import { compose, equals, partial } from 'ramda';
import hash from 'object-hash';

import { mapWithIndex } from 'utils';

import { propTypes, defaultProps } from './Breadcrumbs.props';
import { StyledBreadcrumbs, Crumb, CrumbLink } from './Breadcrumbs.styles';

const renderLinks = ({ pathname }, { label, to, ...props }, i) => (
  <Crumb key={hash({ label: label && label.toString(), to, i })}>
    {to ? (
      <CrumbLink data-active={equals(pathname, to)} to={to} {...props}>
        {label}
      </CrumbLink>
    ) : (
      label
    )}
  </Crumb>
);

export const Breadcrumbs = ({ location: { pathname }, links, ...props }) => (
  <StyledBreadcrumbs {...props}>{mapWithIndex(partial(renderLinks, [{ pathname }]), links)}</StyledBreadcrumbs>
);

Breadcrumbs.propTypes = propTypes;
Breadcrumbs.defaultProps = defaultProps;

export default compose(withRouter)(Breadcrumbs);
