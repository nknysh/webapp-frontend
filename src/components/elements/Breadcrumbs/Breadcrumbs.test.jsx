import React from 'react';

import { Breadcrumbs } from './Breadcrumbs';

const getComponent = props => shallow(<Breadcrumbs location={{ pathname: '' }} {...props} />);

describe('<Breadcrumbs />', () => {
  describe('render', () => {
    it('matches snapshot', () => {
      expect(getComponent()).toMatchSnapshot();
      expect(getComponent({ links: [{ label: <div /> }] })).toMatchSnapshot();
      expect(getComponent({ links: [{ label: 'foo', to: 'bar' }] })).toMatchSnapshot();
      expect(getComponent({ links: [{ label: 'foo', to: 'bar' }, { label: <div /> }] })).toMatchSnapshot();
    });
  });
});
