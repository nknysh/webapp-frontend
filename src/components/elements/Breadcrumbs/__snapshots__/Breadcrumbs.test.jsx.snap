// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Breadcrumbs /> render matches snapshot 1`] = `<Breadcrumbsstyles__StyledBreadcrumbs />`;

exports[`<Breadcrumbs /> render matches snapshot 2`] = `
<Breadcrumbsstyles__StyledBreadcrumbs>
  <Breadcrumbsstyles__Crumb
    key="9ab47eadf44d397992b4cb6b15b1f221e198908b"
  >
    <div />
  </Breadcrumbsstyles__Crumb>
</Breadcrumbsstyles__StyledBreadcrumbs>
`;

exports[`<Breadcrumbs /> render matches snapshot 3`] = `
<Breadcrumbsstyles__StyledBreadcrumbs>
  <Breadcrumbsstyles__Crumb
    key="a82e8c5ff9d56218f85ee5fafe0d30fa7bcbcf3b"
  >
    <Breadcrumbsstyles__CrumbLink
      data-active={false}
      to="bar"
    >
      foo
    </Breadcrumbsstyles__CrumbLink>
  </Breadcrumbsstyles__Crumb>
</Breadcrumbsstyles__StyledBreadcrumbs>
`;

exports[`<Breadcrumbs /> render matches snapshot 4`] = `
<Breadcrumbsstyles__StyledBreadcrumbs>
  <Breadcrumbsstyles__Crumb
    key="a82e8c5ff9d56218f85ee5fafe0d30fa7bcbcf3b"
  >
    <Breadcrumbsstyles__CrumbLink
      data-active={false}
      to="bar"
    >
      foo
    </Breadcrumbsstyles__CrumbLink>
  </Breadcrumbsstyles__Crumb>
  <Breadcrumbsstyles__Crumb
    key="fcea48774eac6847e055e5309259828d20316a02"
  >
    <div />
  </Breadcrumbsstyles__Crumb>
</Breadcrumbsstyles__StyledBreadcrumbs>
`;
